File: clients/python/fylo.py
"""Fylo client — drives the `fylo` binary's persistent NDJSON loop.

No pip dependencies. Requires the `fylo` binary on PATH (brew/scoop) or an
explicit path. One long-lived subprocess keeps the engine warm across calls.

    from fylo import Fylo

    with Fylo("/path/to/db") as db:
        db.create_collection("users")
        doc_id = db.put_data("users", {"name": "Ada", "role": "admin"})
        doc = db.get_latest("users", doc_id)
        admins = db.find_docs("users", {"$ops": [{"role": {"$eq": "admin"}}]})

Each operation method builds the request, sends it, and returns the operation's
`result` (raising FyloError on failure). Method names follow Python's snake_case
convention. `request(op)` remains as a raw escape hatch returning the full
response dict — use it for ops without a dedicated method (branching, schema).
"""

import json
import asyncio
import functools
import inspect
import subprocess
import threading

MAX_REQUEST_BYTES = 1024 * 1024
MAX_RESPONSE_BYTES = 8 * 1024 * 1024


class FyloError(RuntimeError):
    pass


class Fylo:
    def __init__(self, root, binary="fylo"):
        args = [
            binary,
            "exec",
            "--loop",
            "--root",
            root,
            "--max-request-bytes",
            str(MAX_REQUEST_BYTES),
            "--max-response-bytes",
            str(MAX_RESPONSE_BYTES),
        ]
        self._proc = subprocess.Popen(
            args,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            bufsize=0,
        )
        self._lock = threading.Lock()

    def request(self, op):
        """Send one raw machine-protocol op; return the full response dict."""
        line = json.dumps(op, separators=(",", ":")).encode("utf-8")
        if len(line) > MAX_REQUEST_BYTES:
            raise FyloError(f"FYLO request exceeds {MAX_REQUEST_BYTES} bytes")
        with self._lock:  # ponytail: one call in flight; drop the lock only if you pipeline
            if self._proc.poll() is not None:
                raise FyloError("fylo process has exited")
            self._proc.stdin.write(line + b"\n")
            self._proc.stdin.flush()
            reply = self._proc.stdout.readline(MAX_RESPONSE_BYTES + 2)
        if not reply:
            raise FyloError("fylo closed the stream (stderr may have details)")
        if not reply.endswith(b"\n") or len(reply) - 1 > MAX_RESPONSE_BYTES:
            self._proc.kill()
            raise FyloError(f"FYLO response exceeds {MAX_RESPONSE_BYTES} bytes")
        try:
            return json.loads(reply[:-1].decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError) as error:
            self._proc.kill()
            raise FyloError("fylo returned malformed UTF-8 or JSON") from error

    def _op(self, op, **fields):
        payload = {"op": op}
        for key, value in fields.items():
            if value is not None:
                payload[key] = value
        response = self.request(payload)
        if not response.get("ok"):
            raise FyloError((response.get("error") or {}).get("message", "fylo error"))
        return response.get("result")

    # --- Collections ---
    def create_collection(self, collection, kind="document"):
        return self._op("createCollection", collection=collection, kind=kind)

    def drop_collection(self, collection):
        return self._op("dropCollection", collection=collection)

    def inspect_collection(self, collection):
        return self._op("inspectCollection", collection=collection)

    def rebuild_collection(self, collection):
        return self._op("rebuildCollection", collection=collection)

    # --- Durable serverless queue ---
    def queue_publish(self, topic, payload, delay_ms=None, idempotency_key=None):
        return self._op(
            "queuePublish",
            topic=topic,
            payload=payload,
            delayMs=delay_ms,
            idempotencyKey=idempotency_key,
        )

    def queue_claim(
        self, topic, group, max_messages=None, visibility_timeout_ms=None, max_attempts=None
    ):
        return self._op(
            "queueClaim",
            topic=topic,
            group=group,
            maxMessages=max_messages,
            visibilityTimeoutMs=visibility_timeout_ms,
            maxAttempts=max_attempts,
        )

    def queue_ack(self, topic, group, id, receipt):
        return self._op("queueAck", topic=topic, group=group, id=id, receipt=receipt)

    def queue_nack(self, topic, group, id, receipt, delay_ms=None, reason=None):
        return self._op(
            "queueNack",
            topic=topic,
            group=group,
            id=id,
            receipt=receipt,
            delayMs=delay_ms,
            reason=reason,
        )

    def queue_extend(self, topic, group, id, receipt, visibility_timeout_ms=None):
        return self._op(
            "queueExtend",
            topic=topic,
            group=group,
            id=id,
            receipt=receipt,
            visibilityTimeoutMs=visibility_timeout_ms,
        )

    def queue_stats(self, topic, group):
        return self._op("queueStats", topic=topic, group=group)

    def queue_dead_letters(self, topic, group, limit=None):
        return self._op("queueDeadLetters", topic=topic, group=group, limit=limit)

    def queue_process(
        self,
        topic,
        group,
        handler,
        max_messages=1,
        visibility_timeout_ms=30000,
        max_attempts=3,
        retry_delay_ms=0,
    ):
        """Process and settle one bounded batch; handler failures are retried."""
        if not callable(handler):
            raise TypeError("queue handler must be callable")
        deliveries = self.queue_claim(
            topic,
            group,
            max_messages=max_messages,
            visibility_timeout_ms=visibility_timeout_ms,
            max_attempts=max_attempts,
        )
        result = {
            "claimed": len(deliveries),
            "acknowledged": 0,
            "retried": 0,
            "deadLettered": 0,
        }
        for delivery in deliveries:
            failed = False
            try:
                outcome = handler(delivery)
                if inspect.isawaitable(outcome):
                    if inspect.iscoroutine(outcome):
                        outcome.close()
                    raise TypeError("async queue handlers require queue_process_async")
            except Exception:
                failed = True
            if not failed:
                self.queue_ack(topic, group, delivery["id"], delivery["receipt"])
                result["acknowledged"] += 1
            else:
                settled = self.queue_nack(
                    topic,
                    group,
                    delivery["id"],
                    delivery["receipt"],
                    delay_ms=retry_delay_ms,
                    reason="queue handler failed",
                )
                key = "deadLettered" if settled.get("deadLettered") else "retried"
                result[key] += 1
        return result

    def queue_consumer(
        self,
        topic,
        group,
        *,
        max_messages=1,
        visibility_timeout_ms=30000,
        max_attempts=3,
        retry_delay_ms=0,
    ):
        """Decorate a function or method as a one-batch queue invocation."""

        def decorate(handler):
            if not callable(handler):
                raise TypeError("queue consumer can decorate only a callable")

            if inspect.iscoroutinefunction(handler):

                @functools.wraps(handler)
                async def async_consumer(*args, **kwargs):
                    return await self.queue_process_async(
                        topic,
                        group,
                        lambda delivery: handler(*args, delivery, **kwargs),
                        max_messages=max_messages,
                        visibility_timeout_ms=visibility_timeout_ms,
                        max_attempts=max_attempts,
                        retry_delay_ms=retry_delay_ms,
                    )

                async_consumer.__fylo_queue_consumer__ = {
                    "topic": topic,
                    "group": group,
                    "maxMessages": max_messages,
                    "visibilityTimeoutMs": visibility_timeout_ms,
                    "maxAttempts": max_attempts,
                    "retryDelayMs": retry_delay_ms,
                }
                return async_consumer

            @functools.wraps(handler)
            def consumer(*args, **kwargs):
                return self.queue_process(
                    topic,
                    group,
                    lambda delivery: handler(*args, delivery, **kwargs),
                    max_messages=max_messages,
                    visibility_timeout_ms=visibility_timeout_ms,
                    max_attempts=max_attempts,
                    retry_delay_ms=retry_delay_ms,
                )

            consumer.__fylo_queue_consumer__ = {
                "topic": topic,
                "group": group,
                "maxMessages": max_messages,
                "visibilityTimeoutMs": visibility_timeout_ms,
                "maxAttempts": max_attempts,
                "retryDelayMs": retry_delay_ms,
            }
            return consumer

        return decorate

    async def queue_process_async(
        self,
        topic,
        group,
        handler,
        max_messages=1,
        visibility_timeout_ms=30000,
        max_attempts=3,
        retry_delay_ms=0,
    ):
        """Async-handler variant; blocking protocol calls run off the event loop."""
        deliveries = await asyncio.to_thread(
            self.queue_claim,
            topic,
            group,
            max_messages,
            visibility_timeout_ms,
            max_attempts,
        )
        result = {
            "claimed": len(deliveries),
            "acknowledged": 0,
            "retried": 0,
            "deadLettered": 0,
        }
        for delivery in deliveries:
            failed = False
            try:
                await handler(delivery)
            except Exception:
                failed = True
            if not failed:
                await asyncio.to_thread(
                    self.queue_ack, topic, group, delivery["id"], delivery["receipt"]
                )
                result["acknowledged"] += 1
            else:
                settled = await asyncio.to_thread(
                    self.queue_nack,
                    topic,
                    group,
                    delivery["id"],
                    delivery["receipt"],
                    retry_delay_ms,
                    "queue handler failed",
                )
                key = "deadLettered" if settled.get("deadLettered") else "retried"
                result[key] += 1
        return result

    # --- Documents ---
    def put_data(self, collection, data):
        return self._op("putData", collection=collection, data=data)

    def batch_put_data(self, collection, batch):
        return self._op("batchPutData", collection=collection, batch=batch)

    def get_doc(self, collection, id):
        return self._op("getDoc", collection=collection, id=id)

    def get_meta(self, collection, id):
        return self._op("getMeta", collection=collection, id=id)

    def set_meta(self, collection, id, meta):
        return self._op("setMeta", collection=collection, id=id, meta=meta)

    def get_latest(self, collection, id, only_id=False):
        return self._op("getLatest", collection=collection, id=id, onlyId=only_id)

    def patch_doc(self, collection, id, new_doc, old_doc=None):
        return self._op("patchDoc", collection=collection, id=id, newDoc=new_doc, oldDoc=old_doc)

    def patch_docs(self, collection, update):
        return self._op("patchDocs", collection=collection, update=update)

    def del_doc(self, collection, id):
        return self._op("delDoc", collection=collection, id=id)

    def del_docs(self, collection, criteria):
        return self._op("delDocs", collection=collection, delete=criteria)

    def restore_doc(self, collection, id):
        return self._op("restoreDoc", collection=collection, id=id)

    # --- Query ---
    def find_docs(self, collection, query):
        return self._op("findDocs", collection=collection, query=query)

    def find_deleted_docs(self, collection, query=None):
        return self._op("findDeletedDocs", collection=collection, query=query or {})

    def find_docs_page(self, collection, query, page=None):
        return self._op("findDocs", collection=collection, query=query, page=page or {})

    def find_deleted_docs_page(self, collection, query=None, page=None):
        return self._op(
            "findDeletedDocs", collection=collection, query=query or {}, page=page or {}
        )

    def join_docs(self, join):
        return self._op("joinDocs", join=join)

    def execute_sql(self, sql, access=None):
        return self._op("executeSQL", sql=sql, access=access)

    def sql(self, query, access=None):
        """Run raw SQL, built with a native f-string: db.sql(f"... {x}").
        Values are inlined verbatim — escape/validate untrusted input yourself.
        """
        return self.execute_sql(query, access)

    def import_bulk_data(self, collection, url, limit_or_options=None):
        return self._op(
            "importBulkData", collection=collection, url=url, limitOrOptions=limit_or_options
        )

    # Collection-scoped facade with short method names, so
    # `db.collection("users").put(data)` reads like the browser client.
    def collection(self, name):
        return _Collection(self, name)

    def close(self):
        if self._proc.poll() is None:
            self._proc.stdin.close()
            self._proc.wait(timeout=30)

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.close()

    def __getattr__(self, name):
        # Sugar: `db.users.put(...)` -> `db.collection("users").put(...)`. Only
        # fires for names that aren't real attributes; skip private/dunder.
        if name.startswith("_"):
            raise AttributeError(name)
        return _Collection(self, name)


class _Collection:
    """A collection-scoped view; methods drop the leading collection argument."""

    def __init__(self, db, name):
        self._db = db
        self._name = name

    def create(self, kind="document"):
        return self._db.create_collection(self._name, kind)

    def drop(self):
        return self._db.drop_collection(self._name)

    def inspect(self):
        return self._db.inspect_collection(self._name)

    def rebuild(self):
        return self._db.rebuild_collection(self._name)

    def put(self, data):
        return self._db.put_data(self._name, data)

    def get(self, id):
        return self._db.get_doc(self._name, id)

    def get_metadata(self, id):
        return self._db.get_meta(self._name, id)

    def set_metadata(self, id, meta):
        return self._db.set_meta(self._name, id, meta)

    def latest(self, id, only_id=False):
        return self._db.get_latest(self._name, id, only_id)

    def patch(self, id, new_doc, old_doc=None):
        return self._db.patch_doc(self._name, id, new_doc, old_doc)

    def delete(self, id):
        return self._db.del_doc(self._name, id)

    def restore(self, id):
        return self._db.restore_doc(self._name, id)

    def find(self, query):
        return self._db.find_docs(self._name, query)

    def find_page(self, query, page=None):
        return self._db.find_docs_page(self._name, query, page)

File: clients/ruby/fylo.rb
# Fylo client — drives the `fylo` binary's persistent NDJSON loop.
#
# No gems (json + open3 are stdlib). Requires the `fylo` binary on PATH
# (brew/scoop) or an explicit path. One long-lived subprocess keeps the engine
# warm across calls.
#
#   require_relative "fylo"
#
#   Fylo.open("/path/to/db") do |db|
#     db.create_collection("users")
#     id = db.put_data("users", { "name" => "Ada", "role" => "admin" })
#     doc = db.get_latest("users", id)
#     admins = db.find_docs("users", { "$ops" => [{ "role" => { "$eq" => "admin" } }] })
#   end
#
# Each operation method builds the request and returns the op's `result`
# (raising FyloError on failure). Method names follow Ruby's snake_case
# convention; object arguments are native Hashes. `request(op)` is the raw
# escape hatch returning the full response Hash.

require "json"
require "open3"

MAX_REQUEST_BYTES = 1024 * 1024
MAX_RESPONSE_BYTES = 8 * 1024 * 1024

class FyloError < StandardError; end

class Fylo
  ENVIRONMENT_NAME = /\A[A-Za-z_][A-Za-z0-9_]*\z/

  def self.environment_file(path)
    values = {}
    File.read(path, encoding: "bom|utf-8").each_line.with_index(1) do |original, line|
      entry = original.strip
      next if entry.empty? || entry.start_with?("#")
      entry = entry.sub(/\Aexport\s+/, "")
      separator = entry.index("=")
      raise FyloError, "#{path}:#{line}: expected NAME=VALUE" unless separator
      name = entry[0...separator].strip
      raise FyloError, "#{path}:#{line}: invalid environment variable name #{name.inspect}" unless ENVIRONMENT_NAME.match?(name)
      input = entry[(separator + 1)..].lstrip
      if input.start_with?("\"", "'")
        quote = input[0]
        value = +""
        index = 1
        closed = false
        while index < input.length
          character = input[index]
          if character == quote
            trailing = input[(index + 1)..].strip
            raise FyloError, "#{path}:#{line}: unexpected characters after quoted value" unless trailing.empty? || trailing.start_with?("#")
            closed = true
            break
          end
          if quote == "\"" && character == "\\"
            index += 1
            raise FyloError, "#{path}:#{line}: unterminated escape sequence" if index >= input.length
            escaped = input[index]
            value << ({ "n" => "\n", "r" => "\r", "t" => "\t", "\"" => "\"", "\\" => "\\" }[escaped] || "\\#{escaped}")
          else
            value << character
          end
          index += 1
        end
        raise FyloError, "#{path}:#{line}: unterminated quoted value" unless closed
        values[name] = value
      else
        values[name] = input.split("#", 2).first.rstrip
      end
    end
    values
  rescue SystemCallError, EncodingError => error
    raise FyloError, "cannot read FYLO environment file #{path}: #{error.message}"
  end

  def self.child_environment(config)
    return nil if config.nil?
    path = config.respond_to?(:to_path) ? config.to_path : config.to_s
    overrides = config.is_a?(Hash) ? config : environment_file(path)
    windows = /mswin|mingw/i.match?(RUBY_PLATFORM)
    environment = ENV.to_h.each_with_object({}) do |(name, value), result|
      result[windows ? name.upcase : name] = value
    end
    overrides.each do |name, value|
      name = name.to_s
      raise ArgumentError, "invalid environment variable name #{name.inspect}" if name.empty? || name.include?("=") || name.include?("\0")
      environment_name = windows ? name.upcase : name
      if value.nil?
        environment.delete(environment_name)
      else
        environment_value = value.to_s
        if environment_value.include?("\0")
          raise ArgumentError, "invalid environment variable value for #{name.inspect}: NUL is not allowed"
        end
        environment[environment_name] = environment_value
      end
    end
    environment
  end

  def self.open(root, binary: "fylo", env: nil)
    db = new(root, binary: binary, env: env)
    return db unless block_given?
    begin
      yield db
    ensure
      db.close
    end
  end

  def initialize(root, binary: "fylo", env: nil)
    args = [
      binary, "exec", "--loop", "--root", root,
      "--max-request-bytes", MAX_REQUEST_BYTES.to_s,
      "--max-response-bytes", MAX_RESPONSE_BYTES.to_s
    ]
    environment = self.class.child_environment(env)
    @stdin, @stdout, @wait = if environment
      Open3.popen2(environment, *args, unsetenv_others: true)
    else
      Open3.popen2(*args)
    end
    @mutex = Mutex.new
  end

  # Send one raw machine-protocol op; return the full response Hash.
  def request(op)
    payload = JSON.generate(op)
    raise FyloError, "FYLO request exceeds #{MAX_REQUEST_BYTES} bytes" if payload.bytesize > MAX_REQUEST_BYTES
    reply = @mutex.synchronize do
      raise FyloError, "fylo process has exited" unless @wait.alive?
      @stdin.puts(payload)
      @stdin.flush
      @stdout.gets(MAX_RESPONSE_BYTES + 2)
    end
    raise FyloError, "fylo closed the stream" if reply.nil?
    unless reply.end_with?("\n") && reply.bytesize - 1 <= MAX_RESPONSE_BYTES
      Process.kill("KILL", @wait.pid) rescue nil
      raise FyloError, "FYLO response exceeds #{MAX_RESPONSE_BYTES} bytes"
    end
    begin
      JSON.parse(reply)
    rescue JSON::ParserError, EncodingError => error
      Process.kill("KILL", @wait.pid) rescue nil
      raise FyloError, "fylo returned malformed UTF-8 or JSON: #{error.message}"
    end
  end

  # --- Collections ---
  def create_collection(collection, kind = "document")
    op("createCollection", "collection" => collection, "kind" => kind)
  end

  def drop_collection(collection)
    op("dropCollection", "collection" => collection)
  end

  def inspect_collection(collection)
    op("inspectCollection", "collection" => collection)
  end

  def rebuild_collection(collection)
    op("rebuildCollection", "collection" => collection)
  end

  # --- Durable serverless queue ---
  def queue_publish(topic, payload, delay_ms: nil, idempotency_key: nil)
    op("queuePublish", "topic" => topic, "payload" => payload, "delayMs" => delay_ms, "idempotencyKey" => idempotency_key)
  end

  def queue_claim(topic, group, max_messages: nil, visibility_timeout_ms: nil, max_attempts: nil)
    op("queueClaim", "topic" => topic, "group" => group, "maxMessages" => max_messages, "visibilityTimeoutMs" => visibility_timeout_ms, "maxAttempts" => max_attempts)
  end

  def queue_ack(topic, group, id, receipt)
    op("queueAck", "topic" => topic, "group" => group, "id" => id, "receipt" => receipt)
  end

  def queue_nack(topic, group, id, receipt, delay_ms: nil, reason: nil)
    op("queueNack", "topic" => topic, "group" => group, "id" => id, "receipt" => receipt, "delayMs" => delay_ms, "reason" => reason)
  end

  def queue_extend(topic, group, id, receipt, visibility_timeout_ms: nil)
    op("queueExtend", "topic" => topic, "group" => group, "id" => id, "receipt" => receipt, "visibilityTimeoutMs" => visibility_timeout_ms)
  end

  def queue_stats(topic, group)
    op("queueStats", "topic" => topic, "group" => group)
  end

  def queue_dead_letters(topic, group, limit: nil)
    op("queueDeadLetters", "topic" => topic, "group" => group, "limit" => limit)
  end

  # Process and settle one bounded batch. Ruby has no decorator syntax, so
  # queue_consumer returns an equivalent callable wrapper.
  def queue_process(topic, group, max_messages: 1, visibility_timeout_ms: 30_000,
                    max_attempts: 3, retry_delay_ms: 0, &handler)
    raise ArgumentError, "queue handler block is required" unless handler
    deliveries = queue_claim(topic, group, max_messages: max_messages,
                             visibility_timeout_ms: visibility_timeout_ms,
                             max_attempts: max_attempts)
    result = { "claimed" => deliveries.length, "acknowledged" => 0,
               "retried" => 0, "deadLettered" => 0 }
    deliveries.each do |delivery|
      failed = false
      begin
        handler.call(delivery)
      rescue StandardError
        failed = true
      end
      unless failed
        queue_ack(topic, group, delivery.fetch("id"), delivery.fetch("receipt"))
        result["acknowledged"] += 1
      else
        settled = queue_nack(topic, group, delivery.fetch("id"), delivery.fetch("receipt"),
                             delay_ms: retry_delay_ms,
                             reason: "queue handler failed")
        result[settled["deadLettered"] ? "deadLettered" : "retried"] += 1
      end
    end
    result
  end

  def queue_consumer(topic, group, **options, &handler)
    raise ArgumentError, "queue handler block is required" unless handler
    lambda { queue_process(topic, group, **options, &handler) }
  end

  # --- Documents ---
  def put_data(collection, data)
    op("putData", "collection" => collection, "data" => data)
  end

  def batch_put_data(collection, batch)
    op("batchPutData", "collection" => collection, "batch" => batch)
  end

  def get_doc(collection, id)
    op("getDoc", "collection" => collection, "id" => id)
  end

  def get_meta(collection, id)
    op("getMeta", "collection" => collection, "id" => id)
  end

  def set_meta(collection, id, meta)
    op("setMeta", "collection" => collection, "id" => id, "meta" => meta)
  end

  def get_latest(collection, id)
    op("getLatest", "collection" => collection, "id" => id)
  end

  def patch_doc(collection, id, new_doc)
    op("patchDoc", "collection" => collection, "id" => id, "newDoc" => new_doc)
  end

  def patch_docs(collection, update)
    op("patchDocs", "collection" => collection, "update" => update)
  end

  def del_doc(collection, id)
    op("delDoc", "collection" => collection, "id" => id)
  end

  def del_docs(collection, criteria)
    op("delDocs", "collection" => collection, "delete" => criteria)
  end

  def restore_doc(collection, id)
    op("restoreDoc", "collection" => collection, "id" => id)
  end

  # --- Query ---
  def find_docs(collection, query)
    op("findDocs", "collection" => collection, "query" => query)
  end

  def find_deleted_docs(collection, query = {})
    op("findDeletedDocs", "collection" => collection, "query" => query)
  end

  def find_docs_page(collection, query, page = {})
    op("findDocs", "collection" => collection, "query" => query, "page" => page)
  end

  def find_deleted_docs_page(collection, query = {}, page = {})
    op("findDeletedDocs", "collection" => collection, "query" => query, "page" => page)
  end

  def join_docs(join)
    op("joinDocs", "join" => join)
  end

  def execute_sql(sql, access = nil)
    op("executeSQL", "sql" => sql, "access" => access)
  end

  # Run raw SQL, built with native interpolation: db.sql("... #{x}").
  # Values are inlined verbatim — escape/validate untrusted input yourself.
  def sql(query, access = nil)
    execute_sql(query, access)
  end

  def import_bulk_data(collection, url)
    op("importBulkData", "collection" => collection, "url" => url)
  end

  def close
    return unless @wait.alive?
    @stdin.close # EOF ends the loop
    @wait.value
  end

  # Collection-scoped facade with short method names, so
  # `db.collection("users").put(data)` reads like the browser client.
  def collection(name)
    Collection.new(self, name)
  end

  CONVERSIONS = %i[to_ary to_hash to_str to_int to_a to_proc to_io].freeze
  private_constant :CONVERSIONS

  # Sugar: `db.users.put(...)` -> `db.collection("users").put(...)`.
  def method_missing(name, *args, &block)
    return super if block || !args.empty? || CONVERSIONS.include?(name) ||
                    name.to_s.end_with?("=", "?", "!")
    collection(name.to_s)
  end

  def respond_to_missing?(name, include_private = false)
    return super if CONVERSIONS.include?(name) || name.to_s.end_with?("=", "?", "!")
    true
  end

  private

  def op(name, fields)
    payload = { "op" => name }
    fields.each { |k, v| payload[k] = v unless v.nil? }
    resp = request(payload)
    raise FyloError, (resp.dig("error", "message") || "fylo error") unless resp["ok"]
    resp["result"]
  end
end

# A collection-scoped view; methods drop the leading collection argument.
class Fylo
  class Collection
    def initialize(db, name)
      @db = db
      @name = name
    end

    def create(kind = "document")
      @db.create_collection(@name, kind)
    end

    def drop
      @db.drop_collection(@name)
    end

    # NB: Object#inspect is used by irb/p, so it stays a safe repr here — call
    # `db.inspect_collection(name)` for the collection's metadata.
    def inspect
      "#<Fylo::Collection #{@name}>"
    end

    def rebuild
      @db.rebuild_collection(@name)
    end

    def put(data)
      @db.put_data(@name, data)
    end

    def get(id)
      @db.get_doc(@name, id)
    end

    def get_metadata(id)
      @db.get_meta(@name, id)
    end

    def set_metadata(id, meta)
      @db.set_meta(@name, id, meta)
    end

    def latest(id)
      @db.get_latest(@name, id)
    end

    def patch(id, new_doc)
      @db.patch_doc(@name, id, new_doc)
    end

    def delete(id)
      @db.del_doc(@name, id)
    end

    def restore(id)
      @db.restore_doc(@name, id)
    end

    def find(query)
      @db.find_docs(@name, query)
    end

    def find_page(query, page = {})
      @db.find_docs_page(@name, query, page)
    end
  end
end
